miio/gateway.py

```diff
@@ -529,45 +581,295 @@ def set_light(self, color_name, brightness):
         brightness_and_color = brightness_and_color_to_int(
             brightness, color_map[color_name]
         )
-        return self.send("set_rgb", [brightness_and_color])
+        return self._gateway.send("set_rgb", [brightness_and_color])
 
 
-class SubDevice:
-    def __init__(self, gw, sid, type_, _, __, ___):
-        self.gw = gw
-        self.sid = sid
-        self.type = DeviceType(type_)
+class SubDevice(Device):
+    """
+    Base class for all subdevices of the gateway
+    these devices are connected through zigbee.
+    """
 
-    def unpair(self):
-        return self.gw.send("remove_device", [self.sid])
+    def __init__(
+        self,
+        gw: Gateway = None,
+        dev_info: SubDeviceInfo = None,
+        ip: str = None,
+        token: str = None,
+        start_id: int = 0,
+        debug: int = 0,
+        lazy_discover: bool = True,
+    ) -> None:
+        if gw is not None:
+            self._gw = gw
+        else:
+            self._gw = Device(ip, token, start_id, debug, lazy_discover)
+            _LOGGER.debug(
+                "Creating new device instance, only use this for cli interface"
+            )
 
-    def battery(self):
-        return self.gw.send("get_battery", [self.sid])[0]
+        if dev_info is None:
+            raise Exception("sid of the subdevice needs to be specified")
 
-    def get_firmware_version(self) -> Optional[int]:
-        """Returns firmware version"""
+        self._gw = gw
+        self.sid = dev_info.sid
+        self._battery = None
         try:
-            return self.gw.send("get_device_prop", [self.sid, "fw_ver"])[0]
-        except Exception as ex:
-            _LOGGER.debug(
-                "Got an exception while fetching fw_ver: %s", ex, exc_info=True
-            )
-            return None
+            self.type = DeviceType(dev_info.type_id)
+        except ValueError:
+            self.type = DeviceType(-1)
 
     def __repr__(self):
         return "<Subdevice %s: %s fw: %s bat: %s>" % (
-            self.type,
+            self.device_type,
             self.sid,
             self.get_firmware_version(),
```

Comment:
I just tested this PR on my gateway (lumi.gateway.v3, firmware 1.4.1_150) does not seem to support this command which makes it crash when doing `discover_devices` via the cli:
```
DEBUG:miio.miioprotocol:192.168.xx.xx:54321 >>: {'id': 5, 'method': 'get_device_prop', 'params': ['lumi.158d00015694a4', 'fw_ver']}
DEBUG:miio.miioprotocol:192.168.xx.xx:54321 (ts: 1970-01-01 00:08:52, id: 5) << {'id': 5, 'error': {'code': -30001, 'message': 'Resp invalid json.'}}
ERROR:miio.miioprotocol:Got error when receiving: {'code': -30001, 'message': 'Resp invalid json.'}
DEBUG:miio.click_common:Exception: Got an exception while fetching property fw_ver
```
with all devices (SensorHT, Plug, Motion, Magnet).

Reply:
`get_firmware_version()` should probably catch the error, and return None if it's not available as the return value is already defined as `Optional[int]`?

Reply:
I have firmware version 1.4.1_175 for my gateway, with model lumi.gateway.v3-MW300.
There it does work, regardless I will fix this to use the value from the discovery if this command fails

Reply:
done